File: Problem3.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        if(root == null)
            return new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();
        HashMap<Integer, List<Integer>> map = new HashMap<>();
        Queue<TreeNode> q = new LinkedList<>();
        Queue<Integer> level = new LinkedList<>();
        q.add(root);
        level.add(0);
        int min = 0, max = 0;
        while(!q.isEmpty()){
            TreeNode curr = q.poll();
            int lev = level.poll();
            if(!map.containsKey(lev))
                map.put(lev ,new ArrayList<>());
            map.get(lev).add(curr.val);
            if(curr.left!=null){
                q.add(curr.left);
                level.add(lev-1);
                min = Math.min(min,lev-1);
            }
            if(curr.right!=null){
                q.add(curr.right);
                level.add(lev+1);
                max = Math.max(max,lev+1);
            }
        }
        for(int i = min; i <= max; i++){
            res.add(map.get(i));
        }
        return res;
    }
}

File: Problem3.py
#Time Complexity :O(n)
#Space Complexity :O(n)
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this : No

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        q=deque()
        q.append(root)
        result=""
        while q:
            curr=q.popleft()
            if curr!=None:
                result+=str(curr.val)
                q.append(curr.left)
                q.append(curr.right)
            else:
                result+='#'
            result+=','    
        return result


    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if len(data)==0:
            return None
        splitArr=data.split(',')
        idx=1
        root=TreeNode(int(splitArr[0]))
        q=deque()
        q.append(root)
        while idx<len(splitArr) and q:
            curr=q.popleft()
            if splitArr[idx]!='#':
                curr.left=TreeNode(int(splitArr[idx]))
                q.append(curr.left)
            idx+=1

            if splitArr[idx]!='#':
                curr.right=TreeNode(int(splitArr[idx]))
                q.append(curr.right)
            idx+=1
        return root


# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))